src/toil/wdl/wdltoil.py

```diff
@@ -572,12 +600,22 @@ def evaluate_defaultable_decl(node: WDL.Tree.Decl, environment: WDLBindings, std
     If the name of the declaration is already defined in the environment, return its value. Otherwise, return the evaluated expression.
     """
 
-    if node.name in environment:
-        logger.debug('Name %s is already defined, not using default', node.name)
-        return environment[node.name]
-    else:
-        logger.info('Defaulting %s to %s', node.name, node.expr)
-        return evaluate_decl(node, environment, stdlib)
+    try:
+        if node.name in environment and not isinstance(environment[node.name], WDL.Value.Null):
+            logger.debug('Name %s is already defined with a non-null value, not using default', node.name)
+            return environment[node.name]
+        else:
+            if node.type is not None and not node.type.optional and node.expr is None:
+                # We need a value for this but there isn't one.
+                raise WDL.Error.EvalError(node, f"Value for {node.name} was not provided and no default value is available")
+            logger.info('Defaulting %s to %s', node.name, node.expr)
+            return evaluate_decl(node, environment, stdlib)
+    except Exception:
+        # If something goes wrong, dump.
+        logger.exception("Evaluation failed for %s", node)
+        logger.info("Statement was evaluated in environment:")
```

Comment:
Is this log message incomplete?  Also... should it be `logger.exception` as well?

Reply:
These should be followed by a call to another function that logs the environment. I think I can maybe change that to take a logging function?